File: java/src/net/razorvine/pyro/serializer/PyroExceptionPickler.java
package net.razorvine.pyro.serializer;

import java.io.IOException;
import java.io.OutputStream;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import net.razorvine.pickle.IObjectPickler;
import net.razorvine.pickle.Opcodes;
import net.razorvine.pickle.PickleException;
import net.razorvine.pickle.Pickler;
import net.razorvine.pyro.PyroException;
import net.razorvine.serpent.IClassSerializer;

/**
 * Pickler extension to be able to pickle PyroException objects.
 *  
 * @author Irmen de Jong (irmen@razorvine.net)
 */
public class PyroExceptionPickler implements IObjectPickler, IClassSerializer {

	public void pickle(Object o, OutputStream out, Pickler currentPickler) throws PickleException, IOException {
		PyroException error = (PyroException) o;
		out.write(Opcodes.GLOBAL);
		byte[] output="Pyro4.errors\nPyroError\n".getBytes();
		out.write(output,0,output.length);
		Object[] args = new Object[] { error.getMessage() };
		currentPickler.save(args);
		out.write(Opcodes.REDUCE);
		if(error._pyroTraceback!=null)
		{
			// add _pyroTraceback attribute to the output
			Map<String, Object> tb = new HashMap<String, Object>();
			tb.put("_pyroTraceback", new String[]{ error._pyroTraceback });		// transform single string back into list
			currentPickler.save(tb);
			out.write(Opcodes.BUILD);
		}		
	}

	@SuppressWarnings("unchecked")
	public static Object FromSerpentDict(Map<Object, Object> dict) {
		Object[] args = (Object[]) dict.get("args");
		PyroException ex = new PyroException((String)args[0]);
		Map<String, Object> attrs = (Map<String, Object>)dict.get("attributes");
		// we can only deal with a possible _pyroTraceback attribute
		if(attrs.containsKey("_pyroTraceback"))
		{
			Object tb = attrs.get("_pyroTraceback");
			// if the traceback is a list of strings, create one string from it
			if(tb instanceof List) {
				StringBuilder sb=new StringBuilder();
				for(Object line: (List<?>)tb) {
					sb.append(line);
				}	
				ex._pyroTraceback=sb.toString();
			} else {
				ex._pyroTraceback=(String)tb;
			}
		}
		return ex;
	}

	public Map<String, Object> convert(Object obj) {
		PyroException ex = (PyroException) obj;
		Map<String, Object> dict = new HashMap<String, Object>();
		// {'attributes':{},'__exception__':True,'args':('hello',),'__class__':'PyroError'}
		dict.put("__class__", "PyroError");
		dict.put("__exception__", true);
		dict.put("args", new Object[] {ex.getMessage()});
		Map<String, Object> attrMap = new HashMap<String,Object>();
		if(ex._pyroTraceback!=null)
			attrMap.put("_pyroTraceback", new String[]{ ex._pyroTraceback} );   // transform single string back into list
		dict.put("attributes", attrMap);
		return dict;
	}
}

File: dotnet/Razorvine.Pyrolite/Pyrolite/Pyro/PyroExceptionPickler.cs
/* part of Pyrolite, by Irmen de Jong (irmen@razorvine.net) */

using System.Collections;
using System.IO;
using System.Text;

using Razorvine.Pickle;
// ReSharper disable InvertIf

namespace Razorvine.Pyro
{

/// <summary>
/// Pickler extension to be able to pickle PyroException objects.
/// </summary>
public class PyroExceptionPickler : IObjectPickler {

	public void pickle(object o, Stream outs, Pickler currentPickler) {
		PyroException error = (PyroException) o;
		outs.WriteByte(Opcodes.GLOBAL);
		var output=Encoding.Default.GetBytes("Pyro4.errors\nPyroError\n");
		outs.Write(output,0,output.Length);
		object[] args = { error.Message };
		currentPickler.save(args);
		outs.WriteByte(Opcodes.REDUCE);
		
		if(!string.IsNullOrEmpty(error._pyroTraceback))
		{
			// add _pyroTraceback attribute to the output
			Hashtable tb = new Hashtable {["_pyroTraceback"] = new[] {error._pyroTraceback}};
			// transform single string back into list
			currentPickler.save(tb);
			outs.WriteByte(Opcodes.BUILD);
		}
	}
	
	public static IDictionary ToSerpentDict(object obj)
	{
		PyroException ex = (PyroException) obj;
		IDictionary dict = new Hashtable();
		// {'attributes':{},'__exception__':True,'args':('hello',),'__class__':'PyroError'}
		dict["__class__"] = "PyroError";
		dict["__exception__"] = true;
		if(ex.Message != null)
			dict["args"] = new object[] {ex.Message};
		else
			dict["args"] = new object[0];
		if(!string.IsNullOrEmpty(ex._pyroTraceback))
			ex.Data["_pyroTraceback"] = new [] { ex._pyroTraceback } ;    	// transform single string back into list
		dict["attributes"] = ex.Data;
		return dict;
	}

	public static object FromSerpentDict(IDictionary dict)
	{
		var args = (object[]) dict["args"];
		
		string pythonExceptionType = (string) dict["__class__"];
		PyroException ex;
		if(args.Length==0)
		{
			ex = string.IsNullOrEmpty(pythonExceptionType) ? new PyroException() : new PyroException("["+pythonExceptionType+"]");
		}
		else
		{
			ex = string.IsNullOrEmpty(pythonExceptionType) ? new PyroException((string)args[0]) : new PyroException($"[{pythonExceptionType}] {args[0]}");
		}
		
		ex.PythonExceptionType = pythonExceptionType;

		IDictionary attrs = (IDictionary)dict["attributes"];
		foreach(DictionaryEntry entry in attrs)
		{
			string key = (string)entry.Key;
			ex.Data[key] = entry.Value;
			if("_pyroTraceback"==key)
			{
				// if the traceback is a list of strings, create one string from it
				var tbcoll = entry.Value as ICollection;
				if(tbcoll != null) {
					StringBuilder sb=new StringBuilder();
					foreach(object line in tbcoll) {
						sb.Append(line);
					}	
					ex._pyroTraceback=sb.ToString();
				} else {
					ex._pyroTraceback=(string)entry.Value;
				}
			}
		}
		return ex;
	}
}

}
